includes/gateway/class-omise-payment-alipay.php

```diff
@@ -61,102 +61,60 @@ public function init_form_fields() {
 		}
 
 		/**
-		 * @param  int $order_id
+		 * @since  3.3
 		 *
-		 * @see    WC_Payment_Gateway::process_payment( $order_id )
-		 * @see    woocommerce/includes/abstracts/abstract-wc-payment-gateway.php
+		 * @see    Omise_Payment::process_payment( $order_id )
 		 *
-		 * @return array
+	 	 * @param  int $order_id
+		 * @param  WC_Order $order
+		 *
+		 * @return OmiseCharge|OmiseException
 		 */
-		public function process_payment( $order_id ) {
-			if ( ! $order = $this->load_order( $order_id ) ) {
-				wc_add_notice(
-					sprintf(
-						wp_kses(
-							__( 'We cannot process your payment.<br/>Note that nothing wrong by you, this might be from our store issue.<br/><br/>Please feel free to try submit your order again or report our support team that you have found this problem (Your temporary order id is \'%s\')', 'omise' ),
-							array(
-								'br' => array()
-							)
-						),
-						$order_id
-					),
-					'error'
-				);
-				return;
-			}
+		public function charge( $oder_id, $order ) {
+			$metadata = array_merge(
+				apply_filters( 'omise_charge_params_metadata', array(), $order ),
+				array( 'order_id' => $oder_id ) // override order_id as a reference for webhook handlers
+			);
 
-			$order->add_order_note( __( 'Omise: Processing a payment with Alipay solution..', 'omise' ) );
+			return OmiseCharge::create( array(
+				'amount'      => $this->format_amount_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters( 'omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order ),
+				'source'      => array( 'type' => 'alipay' ),
+				'return_uri'  => add_query_arg( 'order_id', $order_id, site_url() . "?wc-api=omise_alipay_callback" ),
+				'metadata'    => $metadata
+			) );
+		}
 
-			try {
-				$metadata = apply_filters( 'omise_charge_params_metadata', array(), $order );
-				$charge = $this->sale( array(
```

Comment:
Continue from https://github.com/omise/omise-woocommerce/pull/101/files#r259255182,
2. Moving this part up, instead of passing all parameters to `sale` method, this `Omise_Payment_Alipay::charge()` method will create a charge by itself, and return OmiseCharge object back to `Omise_Payment::process_payment`